### PR TITLE
Title Screen

### PR DESCRIPTION
**Pull request by PikalaxALT**
the first part of overlay_60

### CODE REVIEW

**Review by red031000 — CHANGES_REQUESTED** [review]
a few things

File: src/title_screen.c [hunk 1]
```diff
@@ -0,0 +1,802 @@
...
+
+static const struct CameraScript sCameraScript_SS[5] = {
+    {{FX32_CONST(105), FX32_CONST(162), FX32_CONST(291)}, 10},
+    {{FX32_CONST(395), FX32_CONST(432), FX32_CONST(191)}, 5 },
+    {{FX32_CONST(105), FX32_CONST(162), FX32_CONST(291)}, 5 },
+    {{FX32_CONST(420), FX32_CONST(87), FX32_CONST(331)},  10},
+    {{FX32_CONST(0), FX32_CONST(0), FX32_CONST(0)},       0 },
+};
+
+static fx32 fx32_abs(fx32 x) {
```
[the remark this answers is not in the document]
Comment (PikalaxALT) [reply src/title_screen.c:721]: i didn't see anything like this defined in sdk

File: src/title_screen.c [hunk 2]
```diff
@@ -0,0 +1,802 @@
...
+
+#if 0
+// Maybe originally intended, but never used
+#define CAMERA_SPEED (animData->cameraSpeed)
+#else
+#define CAMERA_SPEED (5 * FX32_ONE)
+#endif
+
+static void TitleScreenAnim_GetCameraNextPosition(TitleScreenAnimData *animData) {
+#if 0
```
[the remark this answers is not in the document]
Comment (PikalaxALT) [reply src/title_screen.c:733]: I mentioned it in the discord earlier but this is code I wrote to give the player exclusive control over the camera here

File: src/title_screen.c [hunk 3]
```diff
@@ -0,0 +1,802 @@
...
+
+    if (animData->gameVersion == VERSION_HEARTGOLD) {
+        res1 = NARC_titledemo_titledemo_00000003_NCGR;
+        res2 = NARC_titledemo_titledemo_00000004_NCLR;
+    } else {
+        res1 = NARC_titledemo_titledemo_00000001_NCGR;
+        res2 = NARC_titledemo_titledemo_00000002_NCLR;
+    }
+    GfGfxLoader_LoadCharData(NARC_demo_title_titledemo, res1, bgConfig, GF_BG_LYR_SUB_2, 0, 0, FALSE, heapID);
+    GfGfxLoader_GXLoadPal(NARC_demo_title_titledemo, res2, GF_PAL_LOCATION_SUB_BGEXT, (enum GFPalSlotOffset)0x4000, 0, heapID);
```
[the remark this answers is not in the document]
Comment (PikalaxALT) [reply src/title_screen.c:608]: i disagree, i think the enum should be removed

File: src/title_screen.c [hunk 4]
```diff
@@ -0,0 +1,802 @@
...
+                    NNS_G3dAnmObjSetFrame(ppAnmObj[i], 0);
+                }
+            }
+        }
+    }
+}
+
+static void TitleScreenAnimObjs_Run(TitleScreenAnimObject *animObj) {
+    MtxFx33 mtx = {
+        FX32_ONE, 0, 0, 0, FX32_ONE, 0, 0, 0, FX32_ONE,
```
[the remark this answers is not in the document]
Comment (PikalaxALT) [reply src/title_screen.c:376]: tell that to your .clang-format

File: src/title_screen.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (red031000) [reply src/title_screen.c:721]: I mean fair, especially if it's compiled in, I'm just more suprised that it is here and not in like a general fx file

File: src/title_screen.c [hunk 5]
```diff
@@ -0,0 +1,802 @@
...
+
+    if (data->animData.plttData != NULL) {
+        PaletteData_PushTransparentBuffers(data->animData.plttData);
+    }
+
+    DoScheduledBgGpuUpdates(data->bgConfig);
+}
+
+static void TitleScreen_SetGfxBanks(void) {
+    GraphicsBanks graphicsBanks = {
```
[the remark this answers is not in the document]
Comment (PikalaxALT) [reply src/title_screen.c:280]: from a pure style perspective you're right, but this data is copied onto the stack, strongly implying that this is how gf wrote it. that plus the equivalent code in platinum does the exact same thing.

File: src/title_screen.c [hunk 6]
```diff
@@ -0,0 +1,802 @@
...
+            TitleScreenAnim_Run(&data->animData, data->bgConfig, data->heapID);
+        } else {
+            data->animData.enableStartInstructionFlash = TRUE;
+            TitleScreenAnim_Run(&data->animData, data->bgConfig, data->heapID);
+            ++data->timer;
+            if ((gSystem.newKeys & PAD_BUTTON_A) == PAD_BUTTON_A || (gSystem.newKeys & PAD_BUTTON_START) == PAD_BUTTON_START || gSystem.touchNew) {
+                data->exitMode = TITLESCREEN_EXIT_MENU;
+                GF_SndStartFadeOutBGM(0, 60);
+                PlayCry(TITLE_SCREEN_SPECIES, 0);
+                GF_SetVolumeBySeqNo(1, 48);
```
[the remark this answers is not in the document]
Comment (PikalaxALT) [reply src/title_screen.c:163]: feel free to document the sound routines and structs

File: src/title_screen.c [hunk 7]
```diff
@@ -0,0 +1,802 @@
...
+}
+
+static BOOL TitleScreen_Exit(OVY_MANAGER *man, int *state) {
+    TitleScreenOverlayData *data = OverlayManager_GetData(man);
+    HeapID heapID = data->heapID;
+    int exitMode = data->exitMode;
+
+    Main_SetVBlankIntrCB(NULL, NULL);
+    TitleScreen_Delete3DVramMan(data);
+    TitleScreen_DeinitBgs(data);
```
[the remark this answers is not in the document]
Comment (PikalaxALT) [reply src/title_screen.c:238]: Please suggest an alternative

File: src/title_screen.c [hunk 8]
```diff
@@ -0,0 +1,802 @@
...
+
+static BOOL TitleScreen_Exec(OVY_MANAGER *man, int *state) {
+    TitleScreenOverlayData *data = OverlayManager_GetData(man);
+    switch (*state) {
+    case TITLESCREEN_MAIN_WAIT_FADE:
+        if (TitleScreenAnim_InitObjectsAndCamera(&data->animData, data->bgConfig, data->heapID) == TRUE) {
+            data->animData.state = 0;
+            data->initialDelay = 30;
+            gSystem.unk70 = 0;
+            *state = (int)TITLESCREEN_MAIN_START_MUSIC;
```
[the remark this answers is not in the document]
Comment (PikalaxALT) [reply src/title_screen.c:142]: compiler throws an error without it. i don't make the rules.

[the remark this answers is not in the document]
Comment (red031000) [reply src/title_screen.c:142]: compiler is dumb, that's ok then

File: src/title_screen.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (red031000) [reply src/title_screen.c:238]: DeleteBGs? UnloadBGs?
those sound better than Deinit to me

File: src/title_screen.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (red031000) [reply src/title_screen.c:163]: fair, considering this out of scope then

File: src/title_screen.c [hunk 9]
```diff
@@ -0,0 +1,802 @@
...
+static fx32 fx32_abs(fx32 x);
+static void TitleScreenAnim_GetCameraNextPosition(TitleScreenAnimData *animData);
+static void TitleScreenAnim_FadeInGameTitleLayer(TitleScreenAnimData *animData);
+
+extern const OVY_MGR_TEMPLATE gApplication_IntroMovie;
+
+const OVY_MGR_TEMPLATE gApplication_TitleScreen = {TitleScreen_Init, TitleScreen_Exec, TitleScreen_Exit, (FSOverlayID)-1};
+
+static BOOL TitleScreen_Init(OVY_MANAGER *man, int *state) {
+    sub_0200FBF4(PM_LCD_TOP, RGB_WHITE);
```
[the remark this answers is not in the document]
Comment (PikalaxALT) [reply src/title_screen.c:108]: the corresponding code in retsam is ||src/system/brightness.c|| if you want to try and adapt that

File: src/title_screen.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (red031000) [reply src/title_screen.c:280]: I know that it probably is how GF wrote it, but also consider
GF is dumb and writes bad code

I'd greatly perfer data to be outside of funcs, if it still matches

[the remark this answers is not in the document]
Comment (red031000) [reply src/title_screen.c:280]: and plat doesn't count here, considering that's copied from GF

File: src/title_screen.c [hunk 10]
```diff
@@ -0,0 +1,802 @@
...
+        break;
+    case TITLESCREEN_GLOW_PAUSE:
+        ++animData->glowTimer;
+        if (animData->glowTimer > 20) {
+            animData->glowState = TITLESCREEN_GLOW_SETUP;
+            animData->glowTimer = 0;
+        }
+        break;
+    }
+    PaletteData_FadePalettesTowardsColorStep(animData->plttData, 2, 0xFF00, 160, animData->glowFadeStep, RGB(12, 12, 12));
```
[the remark this answers is not in the document]
Comment (PikalaxALT) [reply src/title_screen.c:667]: are you talking about transparentBit and opaqueBit? these are masks for the palette slots which don't have informative enums

File: src/title_screen.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (red031000) [reply src/title_screen.c:376]: yeah I understand clang format, normally I'd agree, however, matrices are a bit of an exception, if I could figure out how to code that into clang-format I would, but I don't know how

File: src/title_screen.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (red031000) [reply src/title_screen.c:108]: yeah no thanks, if you say it is indeed RGB then I'll take your word for it

File: src/title_screen.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (red031000) [reply src/title_screen.c:608]: ehh, disagree myself on that, having palette slots directly labelled and made clear is an advantage imo

File: src/title_screen.c [hunk 11]
```diff
@@ -0,0 +1,802 @@
...
+        AddTextPrinterParameterizedWithColorAndSpacing(&animData->window, 0, string, 0, 0, TEXT_SPEED_INSTANT, MAKE_TEXT_COLOR(1, 1, 0), 1, 0, NULL);
+    } else {
+        AddTextPrinterParameterizedWithColorAndSpacing(&animData->window, 0, string, 0, 0, TEXT_SPEED_INSTANT, MAKE_TEXT_COLOR(2, 2, 0), 1, 0, NULL);
+    }
+    String_Delete(string);
+    DestroyMsgData(msgData);
+
+    u16 color1 = RGB(27, 8, 0);   // HeartGold
+    u16 color2 = RGB(0, 28, 31);  // SoulSilver
+    BG_LoadPlttData(3, &color1, sizeof(u16), 0x42);
```
[the remark this answers is not in the document]
Comment (red031000) [reply src/title_screen.c:634]: fair enough, GF code weird but ok

File: src/title_screen.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (PikalaxALT) [reply src/title_screen.c:608]: not really, it's just namespace clutter at this point

File: src/title_screen.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (red031000) [reply src/title_screen.c:667]: ignore me, I just looked at the func itself, only thing is given that 2 is a bit, should probably be 0x02 instead

File: src/title_screen.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (red031000) [reply src/title_screen.c:735]: ok fair, imo should still be in a branch of itself, or maybe #if DEBUG instead

File: asm/overlay_112.s [hunk 12]
```diff
@@ -2,7 +2,7 @@
 	.include "asm/macros.inc"
 	.include "overlay_112.inc"
 	.include "global.inc"
-	.public ov60_021EAFE0
+	.public gApplication_TitleScreen
```
Comment (adrienntindall) [inline asm/overlay_112.s:5]: ideally this should go in the .inc file but it's nbd

File: src/title_screen.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (adrienntindall) [reply src/title_screen.c:735]: yeah right now the project should be focused on decomp, extension is fine but not in the main branch

**Review by adrienntindall — CHANGES_REQUESTED** [review]

File: src/title_screen.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (PikalaxALT) [reply src/title_screen.c:735]: perhaps i should create a define for this: `TITLE_SCREEN_CAMERA_DEBUG` sound ok?

File: src/title_screen.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (red031000) [reply src/title_screen.c:280]: please move data outside of func

**Review by red031000 — DISMISSED** [review]

**Review by red031000 — DISMISSED** [review]

**Review by adrienntindall — DISMISSED** [review]

**Review by adrienntindall — APPROVED** [review]

**Pull request merged**